mixer/adapter/kubernetesenv/kubernetesenv.go

```diff
@@ -138,30 +145,49 @@ func (b *builder) Build(ctx context.Context, env adapter.Env) (adapter.Handler,
 	if !exists {
 		path = paramsProto.KubeconfigPath
 	}
-	clientset, err := b.newClientFn(path, env)
+
+	config, err := b.newConfig(path, env)
 	if err != nil {
-		return nil, err
-	}
-	controller := newCacheController(clientset, refresh)
-	env.ScheduleDaemon(func() { controller.Run(stopChan) })
-	// ensure that any request is only handled after
-	// a sync has occurred
-	env.Logger().Infof("Waiting for kubernetes cache sync...")
-	if success := cache.WaitForCacheSync(stopChan, controller.HasSynced); !success {
-		stopChan <- struct{}{}
-		return nil, errors.New("cache sync failure")
-	}
-	env.Logger().Infof("Cache sync successful.")
+		return nil, fmt.Errorf("could not build config: %v", err)
+	}
+
+	// only ever build a controller for a config once. this potential blocks
+	// the Build() for multiple handlers using the same config until the first
+	// one has synced. This should be OK, as the WaitForCacheSync was meant to
+	// provide this basic functionality before.
+	b.Lock()
+	controller, found := b.controllers[path]
+	if !found {
+		clientset, err := b.newClientset(config)
+		if err != nil {
```

Comment:
b isn't unlocked here. I think you should put `defer b.Unlock()` at at line 159.

Reply:
added defer